packages/react-aria/src/hooks/useARIAButton.ts

```diff
@@ -0,0 +1,116 @@
+
+/** @internal */
+export const _KeyboardEventKeys = {
+  SPACE_BAR: ' ',
+  ENTER: 'Enter',
+} as const;
+
+/**
+ * button keyboard handling, role, disabled and tabIndex implementation that ensures ARIA spec
+ * for multiple scenarios of shorthand properties. Ensuring 1st rule of ARIA for cases
+ * where no attribute addition is required
+ */
+export function useARIAButton(
+  shorthand: ObjectShorthandProps<React.ButtonHTMLAttributes<HTMLElement>>,
+): ObjectShorthandProps<React.ButtonHTMLAttributes<HTMLElement>> {
+  const { onClick, onKeyDown, onKeyUp, disabled: defaultDisabled, ['aria-disabled']: ariaDisabled } = shorthand;
+  const disabled = mergeARIADisabled(defaultDisabled ?? ariaDisabled);
+
+  const onClickHandler = useEventCallback((ev: React.MouseEvent<HTMLElement>) => {
+    if (disabled) {
+      ev.preventDefault();
+      ev.stopPropagation();
+    } else {
+      if (typeof onClick === 'function') {
+        onClick(ev);
+      }
+    }
+  });
+
+  const onKeyDownHandler = useEventCallback((ev: React.KeyboardEvent<HTMLElement>) => {
+    if (typeof onKeyDown === 'function') {
+      onKeyDown(ev);
+    }
+    if (ev.isDefaultPrevented()) {
+      return;
+    }
+    if (disabled && (ev.key === _KeyboardEventKeys.ENTER || ev.key === _KeyboardEventKeys.SPACE_BAR)) {
+      ev.preventDefault();
+      ev.stopPropagation();
+      return;
+    }
+    if (ev.key === _KeyboardEventKeys.SPACE_BAR) {
+      ev.preventDefault();
+      return;
+    }
+    // If enter is pressed, activate the button
+    else if (ev.key === _KeyboardEventKeys.ENTER) {
+      ev.preventDefault();
+      ev.currentTarget.click();
+    }
+  });
+
+  const onKeyupHandler = useEventCallback((ev: React.KeyboardEvent<HTMLElement>) => {
+    if (typeof onKeyUp === 'function') {
+      onKeyUp(ev);
+    }
+    if (ev.isDefaultPrevented()) {
+      return;
+    }
+    if (disabled && (ev.key === _KeyboardEventKeys.ENTER || ev.key === _KeyboardEventKeys.SPACE_BAR)) {
+      ev.preventDefault();
+      ev.stopPropagation();
+      return;
+    }
+    if (ev.key === _KeyboardEventKeys.SPACE_BAR) {
+      ev.preventDefault();
+      ev.currentTarget.click();
+    }
+  });
+
+  if (!shorthand.hasOwnProperty('as') || shorthand.as === 'button') {
+    shorthand.as = 'button';
+    return shorthand; // there's nothing to be done if as prop === 'button'
+  }
+
+  if (!shorthand.hasOwnProperty('children')) {
+    shorthand.children = null;
+  }
+
+  /**
+   * TODO: Ideally this is unnecessary after implementation of as-prop RFC.
+   * The way to go is to have an assertion method to ensure types,
+   * in the case of button we'd like to limit it for: button, div, span, a
+   */
+  if (typeof shorthand.as !== 'string') {
+    return shorthand;
+  }
+
+  if (!shorthand.hasOwnProperty('role')) {
+    shorthand.role = 'button';
+  }
+  if (!shorthand.hasOwnProperty('aria-disabled')) {
+    shorthand['aria-disabled'] = disabled;
```

Comment:
does this mean that every element will get `aria-disabled` applied ? we should only apply `aria-disabled` when native disabled `disabled` can't be applied

Reply:
As you can see in the [test file](https://github.com/microsoft/fluentui/blob/9723d536c188b95524e814f541db5d77bcfc1b18/packages/react-aria/src/hooks/useARIAButton.test.tsx#L13) in the case of `button` this doesn't happen. and there's no other case where native disabled can be applied as far as I know....

Reply:
For what I understood on the discussion of 1st rule of ARIA, things like button should support a limited amount of elements, like: `div`, `span`. `a` and that's it!